phases/register.go

```diff
@@ -1,9 +1,119 @@
 package phases
 
-import "fmt"
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
 
-// Register ...
-func Register(progress Progress) error {
-	fmt.Println("Register")
+	"github.com/bitrise-io/go-utils/log"
+)
+
+var baseURL = "https://app.bitrise.io"
+
+var (
+	webhookAttemptCount int
+	webhookAttemptMax   = 3
+)
+
+func startAppRegistration() (string, error) {
+	return "", nil
+}
+
+func registerSSHKey() error {
+	return nil
+}
+
+func registerWebhook(appSlug string, apiToken string) error {
+	webhookAttemptCount = 0
+	url := fmt.Sprintf("%s/app/%s/register-webhook.json", baseURL, appSlug)
+	request, err := http.NewRequest(http.MethodPost, url, nil)
+	if err != nil {
+		return fmt.Errorf("create POST %s request: %s", url, err)
+	}
+
+	request.Header.Add("Authorization", "token "+apiToken)
+	resp, err := http.DefaultClient.Do(request)
+	if err != nil {
+		return fmt.Errorf("send POST %s request: %s", url, err)
+	}
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			log.Errorf("Failed to close response body, error: %s", err)
+		}
+	}()
+
+	if resp.StatusCode == http.StatusBadRequest {
+		log.Errorf("Error registering webhook")
+		log.Warnf("Make sure you have the required access rights to the repository and that you enabled git provider integration jfor your Bitrise account!")
+		log.Warnf("Please fix your configuration and hit enter to try again!")
+
+		reader := bufio.NewReader(os.Stdin)
+
+		for {
+			if webhookAttemptCount == webhookAttemptMax {
+				return fmt.Errorf("maximum number of retries reached")
+			}
+			if _, err = reader.ReadString('\n'); err != nil {
+				log.Errorf("Error reading user input")
+				webhookAttemptCount++
+				continue
+			}
+			if err := registerWebhook(appSlug, apiToken); err != nil {
+				log.Errorf("Error registering webhook: %s", err)
+				log.Warnf("Please fix your configuration and hit enter to try again!")
+				continue
+			}
+			return nil
+		}
+	}
+
+	if resp.StatusCode == http.StatusOK {
```

Comment:
move this block as the first exit phase, currently you have: bad case, ok case, bad case. Make it ok case, bad case, bad case

Reply:
After refactoring the recursion, it seemed sensible to leave this like it is. We iterate while status is bad request, then we check if response is ok, all else fails. If it was ok for the first time, iteration doesn't start.